### PR TITLE
Composition in the universe

### LINKED ISSUE #106: Add HComU
😨

### PR DESCRIPTION
**Pull request by jonsterling**
To resolve #106 

- [x] set up values of hcom-in-universe
- [x] add composite type (the collection of elements of `fhcom`)
- [x] add `box` constructor
- [x] add `cap` to syntax
- [x] add `cap` cuts
- [x] `do_rigid_cap` stuff
- [x] quotation of `cap` cuts
- [x] quotation of hcom type
- [x] quotation in hcom type
- [x] conversion of hcom type
- [x] conversion in hcom type
- [x] conversion of cap cuts
- [x] implement hcom-in-fhcom in `TermBuilder`
- [x] create "destructor" to inspect an hcom-u (like the `pi/sg` destructors), needed for `coe`
- [x] implement coe-in-fhcom in `TermBuilder`
- [x] hook up the above into the `enact_rigid_hcom`
- [x] hook up the above into the `enact_rigid_coe`

I would like help with this PR; it is pretty daunting. One thing to keep in mind is, we need to be very careful in remembering the difference between the type `univ` and the type `el(code-univ)`. That distinction will be the source of many bugs that we can hopefully avoid creating.

### CODE REVIEW

File: src/lib/TermBuilder.ml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 open CoolBasis
 
+exception Todo
```
Comment (favonia) [inline src/lib/TermBuilder.ml:3]: Did you forget something????

File: src/lib/DomainData.ml [hunk 2]
```diff
@@ -65,12 +65,14 @@ and tp =
...
   | Nat
   | TpAbort
+  | TpHCom of dim * dim * cof * con
 
 and hd =
   | Global of Symbol.t
   | Var of int (* De Bruijn level *)
   | Coe of con * dim * dim * con
   | HCom of cut * dim * dim * cof * con
+  (* C*HM: add Cap of dim * dim * cof * con * cut *)
```
Comment (jonsterling) [inline src/lib/DomainData.ml:75]: If anyone is bored and wants to write some code for fun, try adding this constructor and following the rest of the `C*HM` trail 😎

**Comment by jonsterling** [comment]
I am relinquishing this PR to @cangiuli and @favonia; there's just one hard part left, but I'm mentally exhausted and need to take a break from coding.

**Comment by cangiuli** [comment]
(In the note, there is a coherence problem with the isomorphism between el(\widehat{Hcom}) and Hcom.)

**Comment by jonsterling** [comment]
@cangiuli To clarify, the note does not have a coherence problem, but the implementation does.

File: doc/kan.tex [hunk 3]
```diff
@@ -406,75 +407,81 @@ \section{Coercion in composite types}
...
+        \tilde{N} : \Prod{i:\DIM}{
+          \Prod{j:\DIM}{
+            \Prod{*:\TpPrf{\forall i.\phi_i}}{
+              \TpEl{\Code{A}_{i}\prn{j,*}}
+            }
+          }
+        }\\
         \tilde{N}\prn{i,j,*} :=
-          \Coe{\lambda j.\Code{A}_i\prn{j,*}}{s'_i}{j}{\Coe{\lambda i.\Code{A}_i\prn{s'_i,*}}{r}{i}{M}}
+        \Coe{\lambda j.\Code{A}_i\prn{j,*}}{s'_i}{j}{\Coe{\lambda i.\Code{A}_i\prn{s'_i,*}}{r}{i}{M}}
```
Comment (cangiuli) [inline doc/kan.tex:416]: :woman_dancing:

Comment (jonsterling) [reply doc/kan.tex:416]: 🕺

**Comment by jonsterling** [comment]
As an update I have carefully checked all the adjacency conditions in Carlo's rendering of the coe-in-hcom algorithm. I have not yet checked that everything is well-typed / well-defined.

File: src/lib/ConcreteSyntaxData.ml [hunk 4]
```diff
@@ -55,7 +55,6 @@ and con_ =
   | BotC
   | HCom of con * con * con * con * con
   | HFill of con * con * con * con
-  | AutoHCom of con * con * con * con
```
Comment (favonia) [inline src/lib/ConcreteSyntaxData.ml:58]: I thought you still want to keep this?

Comment (jonsterling) [reply src/lib/ConcreteSyntaxData.ml:58]: The fact that I deleted it means that I am getting rid of it for now (there is some subtlety with El that I want to work out)

**Comment by jonsterling** [comment]
OK! I have just a few more changes I want to make.

**Pull request merged**